File: _includes/code/java-v6/src/test/java/BackupsTest.java
import io.weaviate.client6.v1.api.Authentication;
import io.weaviate.client6.v1.api.WeaviateClient;
import io.weaviate.client6.v1.api.backup.Backup;
import io.weaviate.client6.v1.api.backup.BackupStatus;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.MethodOrderer;
import org.junit.jupiter.api.Order;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestMethodOrder;

import java.io.IOException;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.TimeoutException;

import static org.assertj.core.api.Assertions.assertThat;

@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
class BackupsTest {

  private static WeaviateClient client;

  // TODO[g-despot] DX: Authentication.apiKey is not needed, why not string?
  @BeforeAll
  public static void beforeAll() {
    // Connect to the Weaviate instance specified in the Python script
    client = WeaviateClient.connectToLocal(config -> config.port(8580)
        .grpcPort(50551)
        .authentication(Authentication.apiKey("root-user-key")));
  }

  @AfterAll
  public static void afterAll() throws Exception {
    if (client != null) {
      client.close();
    }
  }

  // Helper method to set up collections for tests
  private void setupCollections() throws IOException {
    client.collections.delete("Article");
    client.collections.delete("Publication");
    var articles = client.collections.create("Article");
    var publications = client.collections.create("Publication");

    articles.data.insert(Map.of("title", "Dummy"));
    publications.data.insert(Map.of("title", "Dummy"));
  }

  @Test
  @Order(1)
  void testBackupAndRestoreLifecycle() throws IOException, TimeoutException {
    setupCollections();
    String backupId = "my-very-first-backup";
    String backend = "filesystem";

    // START CreateBackup
    var createResult = client.backup
        .create(backupId, backend,
            backup -> backup.includeCollections("Article", "Publication"))
        .waitForCompletion(client); // Replicates wait_for_completion=True

    System.out.println(createResult);
    // END CreateBackup

    assertThat(createResult.status()).isEqualTo(BackupStatus.SUCCESS);

    // START StatusCreateBackup
    Optional<Backup> createStatus =
        client.backup.getCreateStatus(backupId, backend);

    System.out.println(createStatus.orElse(null));
    // END StatusCreateBackup

    assertThat(createStatus).isPresent();
    assertThat(createStatus.get().status()).isEqualTo(BackupStatus.SUCCESS);

    // Delete all classes before restoring
    client.collections.deleteAll();
    assertThat(client.collections.exists("Article")).isFalse();
    assertThat(client.collections.exists("Publication")).isFalse();

    // START RestoreBackup
    var restoreResult = client.backup.restore(backupId, backend,
        restore -> restore.excludeCollections("Article")
    // Note: roles_restore and users_restore syntax not provided
    ).waitForCompletion(client); // Replicates wait_for_completion=True

    System.out.println(restoreResult);
    // END RestoreBackup

    assertThat(restoreResult.status()).isEqualTo(BackupStatus.SUCCESS);
    // Verify that Publication was restored and Article was excluded
    assertThat(client.collections.exists("Publication")).isTrue();
    assertThat(client.collections.exists("Article")).isFalse();

    // START StatusRestoreBackup
    Optional<Backup> restoreStatus =
        client.backup.getRestoreStatus(backupId, backend);

    System.out.println(restoreStatus.orElse(null));
    // END StatusRestoreBackup

    assertThat(restoreStatus).isPresent();
    assertThat(restoreStatus.get().status()).isEqualTo(BackupStatus.SUCCESS);

    // Clean up
    client.collections.delete("Publication");
  }

  @Test
  @Order(2)
  void testCancelBackup() throws IOException, TimeoutException, InterruptedException {
    setupCollections();
    String backupId = "some-unwanted-backup";
    String backend = "filesystem";

    // Start a backup to cancel (replicates wait_for_completion=False)
    var backupToCancel = client.backup.create(backupId, backend,
        backup -> backup.includeCollections("Article", "Publication"));

    System.out.println(backupToCancel);

    // START CancelBackup
    // Note: The cancel() method is called on the Backup object
    backupToCancel.cancel(client);
    // END CancelBackup

    // Poll until the backup reaches a terminal state.
    // The cancel is async, so it may still be STARTED briefly after the call.
    BackupStatus status = BackupStatus.STARTED;
    for (int i = 0; i < 20; i++) {
      var s = client.backup.getCreateStatus(backupId, backend);
      if (s.isPresent()) {
        status = s.get().status();
        if (status != BackupStatus.STARTED) break;
      }
      Thread.sleep(500);
    }

    // The backup may complete before the cancel arrives,
    // so accept either CANCELED or SUCCESS.
    assertThat(status).isIn(BackupStatus.CANCELED, BackupStatus.SUCCESS);

    // Clean up
    client.collections.delete("Article");
    client.collections.delete("Publication");
  }

  @Test
  @Order(3)
  void testCancelRestore() throws IOException, TimeoutException, InterruptedException {
    setupCollections();
    String backupId = "backup-to-cancel-restore";
    String backend = "filesystem";

    // Create a backup first
    client.backup.create(backupId, backend,
        backup -> backup.includeCollections("Article", "Publication"))
        .waitForCompletion(client);

    // Delete collections before restoring
    client.collections.deleteAll();

    // Start a restore and cancel it
    var restoreToCancel = client.backup.restore(backupId, backend);

    // START CancelRestore
    client.backup.cancelRestore(backupId, backend);
    // END CancelRestore

    // Poll until the restore reaches a terminal state
    BackupStatus status = BackupStatus.STARTED;
    for (int i = 0; i < 20; i++) {
      var s = client.backup.getRestoreStatus(backupId, backend);
      if (s.isPresent()) {
        status = s.get().status();
        if (status != BackupStatus.STARTED && status != BackupStatus.TRANSFERRING) break;
      }
      Thread.sleep(500);
    }

    // The restore may complete before the cancel arrives
    assertThat(status).isIn(BackupStatus.CANCELED, BackupStatus.SUCCESS);
  }
}

File: _includes/code/csharp/BackupsTest.cs
using System;
using System.Threading;
using System.Threading.Tasks;
using Weaviate.Client;
using Weaviate.Client.Models;
using Xunit;

// Run sequentially to prevent backup conflicts on the filesystem backend
[Collection("Sequential")]
public class BackupsTest : IAsyncLifetime
{
    private WeaviateClient client;
    private readonly BackupBackend _backend = new FilesystemBackend();

    public async Task InitializeAsync()
    {
        client = await Connect.Local(restPort: 8580, grpcPort: 50551, credentials: "root-user-key");

        // Ensure a clean state
        await CleanupCollections();
    }

    public Task DisposeAsync()
    {
        // The C# client manages connections automatically.
        return Task.CompletedTask;
    }

    // Helper method to set up collections for tests
    private async Task SetupCollections()
    {
        await CleanupCollections();

        await client.Collections.Create(
            new CollectionCreateParams { Name = "Article", Properties = [Property.Text("title")] }
        );

        await client.Collections.Create(
            new CollectionCreateParams
            {
                Name = "Publication",
                Properties = [Property.Text("title")],
            }
        );

        await client.Collections.Use("Article").Data.Insert(new { title = "Dummy" });
        await client.Collections.Use("Publication").Data.Insert(new { title = "Dummy" });
    }

    private async Task CleanupCollections()
    {
        if (await client.Collections.Exists("Article"))
            await client.Collections.Delete("Article");
        if (await client.Collections.Exists("Publication"))
            await client.Collections.Delete("Publication");
    }

    [Fact]
    public async Task TestBackupAndRestoreLifecycle()
    {
        await SetupCollections();
        string backupId = "my-very-first-backup";

        // START CreateBackup
        var createResult = await client.Backup.CreateSync(
            new BackupCreateRequest(
                Id: backupId,
                Backend: _backend,
                IncludeCollections: ["Article", "Publication"]
            )
        );

        Console.WriteLine($"Status: {createResult.Status}");
        // END CreateBackup

        Assert.Equal(BackupStatus.Success, createResult.Status);

        // START StatusCreateBackup
        var createStatus = await client.Backup.GetStatus(_backend, backupId);

        Console.WriteLine($"Backup ID: {createStatus.Id}, Status: {createStatus.Status}");
        // END StatusCreateBackup

        Assert.Equal(BackupStatus.Success, createStatus.Status);

        // Delete all classes before restoring
        await client.Collections.DeleteAll();
        Assert.False(await client.Collections.Exists("Article"));
        Assert.False(await client.Collections.Exists("Publication"));

        // START RestoreBackup
        var restoreResult = await client.Backup.RestoreSync(
            new BackupRestoreRequest(
                Id: backupId,
                Backend: _backend,
                ExcludeCollections: ["Article"] // Exclude Article from restoration
            )
        );

        Console.WriteLine($"Restore Status: {restoreResult.Status}");
        // END RestoreBackup

        Assert.Equal(BackupStatus.Success, restoreResult.Status);

        // Verify that Publication was restored and Article was excluded
        Assert.True(await client.Collections.Exists("Publication"));
        Assert.False(await client.Collections.Exists("Article"));

        // START StatusRestoreBackup
        // Note: In C#, restore status is often tracked via the returned operation or by polling if async.
        // GetRestoreStatus checks the status of a specific restore job.
        // Since we ran RestoreSync, we know it is done.
        // We can inspect the result returned from RestoreSync directly.
        Console.WriteLine($"Restore ID: {restoreResult.Id}, Status: {restoreResult.Status}");
        // END StatusRestoreBackup

        Assert.Equal(BackupStatus.Success, restoreResult.Status);

        // Clean up
        await client.Collections.Delete("Publication");
    }

    [Fact]
    public async Task TestCancelBackup()
    {
        await SetupCollections();
        string backupId = "some-unwanted-backup";

        // Start a backup to cancel (Async, creates the operation but returns immediately)
        CancellationToken cancellationToken = new CancellationToken();
        var backupOperation = await client.Backup.Create(
            new BackupCreateRequest(
                Id: backupId,
                Backend: _backend,
                ExcludeCollections: ["Article", "Publication"]
            ),
            cancellationToken
        );

        Console.WriteLine($"Backup started with ID: {backupOperation.Current.Id}");

        // START CancelBackup
        await backupOperation.Cancel(cancellationToken);
        // END CancelBackup

        // Wait for the cancellation to be processed
        var finalStatus = await backupOperation.WaitForCompletion();

        // Verify status
        Assert.Equal(BackupStatus.Canceled, finalStatus.Status);

        // Clean up
        await client.Collections.Delete("Article");
        await client.Collections.Delete("Publication");
    }
}
